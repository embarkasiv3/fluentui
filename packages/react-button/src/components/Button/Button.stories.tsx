import * as React from 'react';
import { Button } from './Button';
import { ButtonProps } from './Button.types';
import * as classes from './Button.stories.scss';
import { Stack, Text } from 'office-ui-fabric-react';

const ButtonVariants = (props: ButtonProps) => (
  <div className={classes.hStack}>
    <Button {...props} content="Hello, world" icon="X" />
    <Button {...props} primary content="Hello, world" icon="X" />
    <Button {...props} disabled content="Hello, world" icon="X" />
    <Button {...props} primary disabled content="Hello, world" icon="X" />
  </div>
);

export const ButtonCss = () => (
  <Stack gap={20}>
    <Text variant="xLarge">A button can appear round using the `circular` prop.</Text>
    <ButtonVariants circular />

    <Text variant="xLarge">A button can fill the width of its container using the `fluid` prop.</Text>
    <div className={classes.vStack}>
      <Button fluid content="Hello, world" icon="X" />
      <Button fluid primary content="Hello, world" icon="X" />
      <Button fluid disabled content="Hello, world" icon="X" />
      <Button fluid primary disabled content="Hello, world" icon="X" />
    </div>

    <Text variant="xLarge">A button can contain only an icon using the `iconOnly` prop.</Text>
    <ButtonVariants iconOnly />

    <Text variant="xLarge">A button can be both `circular` and `iconOnly`.</Text>
    <ButtonVariants circular iconOnly />

    <Text variant="xLarge">An icon button can format its Icon to appear before or after its content.</Text>
    <div className={classes.vStack}>
      <ButtonVariants iconPosition="before" />
      <ButtonVariants iconPosition="after" />
    </div>

    <Text variant="xLarge">A button can show a loading indicator using the `loading` prop.</Text>
    <ButtonVariants loading />

    <Text variant="xLarge">A button can be sized.</Text>
    <div className={classes.vStack}>
      <ButtonVariants size="smallest" />
      <ButtonVariants size="smaller" />
      <ButtonVariants size="small" />
      <ButtonVariants size="large" />
      <ButtonVariants size="larger" />
      <ButtonVariants size="largest" />
    </div>
  </Stack>
);

export const ButtonTokens = () => (
  <Stack gap={20}>
    <Text variant="xLarge">A button can be colored using inline tokens.</Text>
    <div className={classes.vStack}>
      <Button
        icon="O"
        tokens={{
          padding: '3px 10px',
          fontSize: '12px',
          fontWeight: '600',
          contentGap: '8px',
          borderRadius: '4px',
          borderWidth: '1px',

          background: 'linear-gradient(-180deg,#34d058,#28a745 90%)',
          contentColor: 'white',
          borderColor: 'rgba(27, 31, 35, 0.2)',

          hovered: {
            background: '#269f42 linear-gradient(-180deg,#2fcb53,#269f42 90%)',
            contentColor: 'white',
            borderColor: 'rgba(27, 31, 35, .5)',
          },

          pressed: {
            background: '#279f43',
            borderColor: 'rgba(27, 31, 35, .5)',
            contentColor: 'white',
            transform: 'none',
          },
        }}
        content="Github: Open issue"
      />
      <Button
        tokens={{
          fontFamily: `"Amazon Ember", Arial, sans-serif`,
          fontSize: '13px',
          fontWeight: '400',
          borderRadius: '3px',
          padding: '0 10px',

          background: 'linear-gradient(to bottom,#f7dfa5,#f0c14b)',
          contentColor: 'rgb(17, 17, 17)',
          borderColor: 'rgb(168, 135, 52) rgb(156, 126, 49) rgb(132, 106, 41)',

          hovered: {
            background: 'linear-gradient(to bottom,#f5d78e,#eeb933)',
            contentColor: 'rgb(17, 17, 17)',
            borderColor: '#a88734 #9c7e31 #846a29',
          },

          pressed: {
            background: '#f0c14b',
            contentColor: 'rgb(17, 17, 17)',
            borderColor: '#a88734 #9c7e31 #846a29',
          },
        }}
        content="Amazon: Proceed to checkout"
      />
      <Button
        tokens={{
          borderWidth: '0px',
          padding: '7px 17px',
          borderRadius: '3px',
          fontFamily: '"Netflix Sans", "Helvetica Neue", Helvetica, Arial, sans-serif',
          fontSize: '16px',
          fontWeight: '400',

          background: 'rgb(229, 9, 20)',
          contentColor: 'rgb(255, 255, 255)',
          borderColor: 'rgb(255, 255, 255)',

          hovered: {
            background: 'rgb(255, 50, 61)',
            borderColor: '#a88734 #9c7e31 #846a29',
          },

          pressed: {
            background: 'rgb(229, 9, 20)',
            borderColor: '#a88734 #9c7e31 #846a29',
          },
        }}
        content="Netflix: Sign In"
      />
      <Button
        tokens={{
          height: '48px',
          borderWidth: '0px',
          padding: '17px 48px',
          borderRadius: '500px',
          fontFamily: 'Circular, Helvetica, Arial, sans-serif',
          fontSize: '14px',
          fontWeight: '700',

          background: 'rgb(29, 185, 84) none repeat scroll 0% 0% / auto padding-box border-box',
          contentColor: 'rgb(255, 255, 255)',
          borderColor: 'rgb(255, 255, 255)',

          hovered: {
            background: 'rgb(30, 215, 96) none repeat scroll 0% 0% / auto padding-box border-box',
            contentColor: 'rgb(255, 255, 255)',
            borderColor: 'rgb(255, 255, 255)',
          },

          pressed: {
            transform: 'none',
            background: '#1aa34a',
            contentColor: 'rgb(255, 255, 255)',
            borderColor: 'rgb(255, 255, 255)',
          },
        }}
        content="Spotify: GET PREMIUM"
      />
      <Button
        tokens={{
          height: '52px',
          background: 'rgba(0, 0, 0, 0) none repeat scroll 0% 0% / auto padding-box border-box',
          contentColor: 'rgb(0, 0, 0)',
          borderColor: 'rgb(0, 0, 0)',
          borderWidth: '2px',
          padding: '17px 48px',
          borderRadius: '500px',
          fontFamily: 'Circular, Helvetica, Arial, sans-serif',
          fontSize: '14px',
          fontWeight: '700',
        }}
        content="Spotify: LEARN MORE"
      />
    </div>
    <Text variant="xLarge">A tokenized button can be customized for any size or padding.</Text>
    <div className={classes.vStack}>
      <ButtonVariants
        tokens={{
          height: '24px',
          fontSize: '12px',
          iconSize: '22px',
          padding: '0 8px',
        }}
      />
      <ButtonVariants
        tokens={{
          height: '70px',
          fontSize: '48px',
          iconSize: '48px',
        }}
      />
    </div>
  </Stack>
);
