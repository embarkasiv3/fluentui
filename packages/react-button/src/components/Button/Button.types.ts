import * as React from 'react';
import { ComponentProps, ShorthandProps } from '@fluentui/react-utilities';
import { ObjectShorthandProps } from '@fluentui/react-utilities';

/**
 * {@docCategory Button}
 */
export type ButtonProps = ComponentProps &
  React.ButtonHTMLAttributes<HTMLElement> & {
    /**
     * Shorthand icon. A shorthand prop can be a literal, object, or
     * JSX. The `children` prop of the object can be a render function,
     * taking in the original slot component and props.
     */
    icon?: ShorthandProps<React.HTMLAttributes<HTMLSpanElement>>;

    // TODO: children needs a new typing to handle render functions along with hook updates for children functionality
    // children?: ???;

    // /** A button can appear circular. */
    // circular?: boolean;

    /** A button can show that it cannot be interacted with. */
    disabled?: boolean;

    // /**
    // eslint-disable-next-line @fluentui/max-len
    //  * When set, allows the button to be focusable even when it has been disabled. This is used in scenarios where it is
    //  * important to keep a consistent tab order for screen reader and keyboard users.
    //  * @defaultvalue false
    //  */
    // disabledFocusable?: boolean;

    // /** A button can fill the width of its container. */
    // block?: boolean;

    /** A button can contain only an icon. */
    iconOnly?: boolean;

    /** An icon button can format its icon to appear before or after its content. */
    iconPosition?: 'before' | 'after';

    // /** A button can show a loading indicator. */
    // loading?: boolean;

    /** A button can emphasize that it represents the primary action. */
    primary?: boolean;

    // /** A button can blend into its background to become less emphasized. */
    // subtle?: boolean;

    // /** A button can have no background styling and just be emphasized through its content styling. */
    // transparent?: boolean;

    /** A button can be sized. */
    size?: 'small' | 'large';
  };

/**
 * {@docCategory Button}
 */
export interface ButtonState extends ButtonProps {
  ref: React.RefObject<HTMLButtonElement>;

  icon?: ObjectShorthandProps<React.HTMLAttributes<HTMLSpanElement>>;
  children?: ObjectShorthandProps<React.HTMLAttributes<HTMLSpanElement>>;
}

/**
 * {@docCategory Button}
 */
export type ButtonStyleSelectors = {
  disabled?: boolean;
  iconOnly?: boolean;
  primary?: boolean;
  size?: string;
};

/**
 * {@docCategory Button}
 */
export type ButtonTokens = {
  height: string;
  maxWidth: string;
  minWidth: string;
  paddingX: string;
  paddingY: string;

  fontSize: string;
  fontWeight: number;
  lineHeight: string;

  iconFontSize: string;
  iconHeight: string;
  iconSpacing: string;
  iconWidth: string;

  background: string;
  color: string;

  borderColor: string;
  borderRadius: string;
  borderWidth: string;

  shadow: string;

  hovered: Partial<{
    background: string;
    borderColor: string;
    color: string;
  }>;

  pressed: Partial<{
    background: string;
    borderColor: string;
    color: string;
    shadow: string;
  }>;
};

/**
 * {@docCategory Button}
 */
export type ButtonVariants =
  | 'base'
  | 'disabled'
  | 'iconOnly'
  | 'primary'
  | 'small'
  | 'large'
  // TODO: get rid of these combinations, use individual variants in matchers
  | 'primaryDisabled'
  | 'iconOnlySmall'
  | 'iconOnlyLarge';

/**
 * {@docCategory Button}
 */
export type ButtonVariantTokens = {
  [variant in ButtonVariants]: Partial<ButtonTokens>;
};
