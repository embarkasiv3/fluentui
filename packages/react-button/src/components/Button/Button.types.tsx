import * as React from 'react';
import { BaseSlots, ComponentProps, SlotProp, SlotProps } from '@fluentui/react-compose';
import { ColorPlateSet } from '@fluentui/react-theme-provider';
import { RecursivePartial } from '../../utils/tempTypes';

export type SizeValue = 'smallest' | 'smaller' | 'small' | 'medium' | 'large' | 'larger' | 'largest';

/**
 * {@docCategory Button}
 */
export interface ButtonRef {
  /**
   * Sets focus to the button.
   */
  focus: () => void;
}

export interface ButtonProps extends ComponentProps, React.HTMLAttributes<HTMLButtonElement> {
  /**
   * Access the imperative API of the Button.
   */
  componentRef?: React.RefObject<ButtonRef>;

  /**
   * Shorthand icon. A shorthand prop can be a literal, object,
   * JSX, or function which takes render options.
   */
  icon?: SlotProp<React.HTMLAttributes<HTMLSpanElement>>;

  /**
   * Shorthand loader content within the button.
   */
  loader?: SlotProp<React.HTMLAttributes<HTMLSpanElement>>;

  /** A button can appear circular. */
  circular?: boolean;

  /** A button can show that it cannot be interacted with. */
  disabled?: boolean;

  /** A button can fill the width of its container. */
  fluid?: boolean;

  /** A button can contain only an icon. */
  iconOnly?: boolean;

  /** An icon button can format its icon to appear before or after its content. */
  iconPosition?: 'before' | 'after';

  /** A button that inherits its background and has a subtle appearance. */
  inverted?: boolean;

  /** A button can show a loading indicator. */
  loading?: boolean;

  /**
   * Called after a user clicks the button.
   * @param event - React's original SyntheticEvent.
   * @param data - All props.
   */
  // onClick?: ComponentEventHandler<ButtonProps>;

  /**
   * Called after a user focuses the button.
   * @param event - React's original SyntheticEvent.
   * @param data - All props.
   */
  // onFocus?: ComponentEventHandler<ButtonProps>;

  /** A button can emphasize that it represents the primary action. */
  primary?: boolean;

  /** A button can emphasize that it represents an alternative action. */
  secondary?: boolean;

  /** A button can be sized. */
  size?: SizeValue;

  // TODO: Deprecate or rename to textOnly for alignment with iconOnly?
  /** A button can be formatted to show only text in order to indicate a less-pronounced action. */
  // text?: boolean;

  tokens?: RecursivePartial<ButtonTokens>;
}

export interface ButtonState extends ButtonProps {
  buttonRef?: React.RefObject<HTMLButtonElement>;
}

export interface ButtonSlots extends BaseSlots {
  icon: React.ElementType;
  loader: React.ElementType;
}

export type ButtonSlotProps = SlotProps<ButtonSlots, ButtonProps, React.ButtonHTMLAttributes<HTMLButtonElement>>;

export type ButtonTokens = ColorPlateSet & {
  /* sizing */
  padding: string;
  margin: string;
  height: string;
  minWidth: string;
  maxWidth: string;
  contentGap: string;
  iconSize: string;
  borderRadius: string;
  borderWidth: string;
  boxShadow: string;
  width: string;

  size: {
    smallest: string;
    smaller: string;
    small: string;
    regular: string;
    large: string;
    larger: string;
    largest: string;
  };

  transform: string;
  transition: string;

  fontFamily: string;
  fontSize: string;
  fontWeight: string;

  pressed: {
    transform: string;
  };
};
