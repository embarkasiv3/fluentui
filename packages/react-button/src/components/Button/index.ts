export * from './Button.types';
export * from './ButtonBase';
export * from './Button';
export * from './useButton';
