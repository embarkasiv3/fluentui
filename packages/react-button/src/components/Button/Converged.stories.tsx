import * as React from 'react';
import { Button } from './Button';
import { mergeThemes, ThemeProvider, Theme } from '@fluentui/react-theme-provider';
import { UploadIcon } from '@fluentui/react-icons';
import { Stack, Text, ColorPicker, IColor } from 'office-ui-fabric-react';

// tslint:disable: jsx-ban-props

const getThemes = (accent: string) => {
  const lightTheme = mergeThemes({
    tokens: {
      palette: {
        accent,
      },

      body: {
        background: 'white',
        contentColor: 'black',
      },

      accent: {
        background: 'var(--palette-accent)',
      },

      button: {
        background: '#ddd',
      },
    },
    stylesheets: [],
  });

  const darkTheme = mergeThemes(lightTheme, {
    tokens: {
      body: {
        background: '#333',
        contentColor: 'white',
      },
      button: {
        background: 'transparent',
        contentColor: 'white',
        hovered: {
          background: '#555',
        },
      },
    },
  });

  return { lightTheme, darkTheme };
};

const ExampleBox = ({ title, theme }: { title: string; theme: Theme }) => (
  <ThemeProvider theme={theme} style={{ padding: 20 }}>
    <Stack
      tokens={{
        childrenGap: 8,
        maxWidth: 400,
      }}
    >
      <Text variant="xLarge">{title}</Text>
      <Button icon={<UploadIcon />} content="I am a button" />
      <Button primary content="I am a primary button" />
    </Stack>
  </ThemeProvider>
);

export const ThemeExample = () => {
  const [color, setColor] = React.useState<string>('#f00');
  const { lightTheme, darkTheme } = getThemes(color);
  const onColorChange = React.useCallback((ev: never, newColor: IColor) => setColor(newColor.str), []);

  return (
    <Stack gap={16}>
      <Text variant="xLarge">Accent color</Text>
      <ColorPicker color={color} onChange={onColorChange} />

      <ExampleBox title="Light" theme={lightTheme} />
      <ExampleBox title="Dark" theme={darkTheme} />
    </Stack>
  );
};
