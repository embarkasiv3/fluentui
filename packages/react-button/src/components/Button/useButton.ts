import { useImperativeHandle, useRef } from 'react';
import { getStyleFromPropsAndOptions } from '@fluentui/react-theme-provider';
import { useFocusRects } from '@uifabric/utilities';
import { ComposePreparedOptions } from '@fluentui/react-compose';
import { ButtonProps, ButtonState } from './Button.types';

/**
 * The useButton hook processes the Button component props and returns state.
 * @param props - Button props to derive state from.
 */
export const useButton = (
  props: ButtonProps,
  ref: React.Ref<HTMLElement>,
  options: ComposePreparedOptions,
): ButtonState => {
  const buttonRef = useRef<HTMLButtonElement | null>(null);

  useImperativeHandle(props.componentRef, () => ({
    focus: () => buttonRef.current?.focus(),
  }));
  useFocusRects(ref as React.RefObject<HTMLElement>);

  return {
    ...props,
    buttonRef,
    style: getStyleFromPropsAndOptions(props, options, '--button'),
  };
};
