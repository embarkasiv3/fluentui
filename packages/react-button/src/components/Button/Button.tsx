import * as React from 'react';
import { useButton } from './useButton';
import { ButtonProps } from './Button.types';
import { useFocusRects } from '@uifabric/utilities';
import { useInlineTokens } from '@fluentui/react-theme-provider';
import { useButtonClasses } from './useButtonClasses';

/**
 * Define a styled Button, using the `useButton` hook.
 */
export const Button = React.forwardRef<HTMLElement, ButtonProps>((props, ref) => {
  const { render, state } = useButton(props, ref);

  // Apply styling.
  useButtonClasses(state);
  useFocusRects(state.ref);
  useInlineTokens(state, '--button');

  // Render component.
  return render(state);
});

Button.displayName = 'Button';
