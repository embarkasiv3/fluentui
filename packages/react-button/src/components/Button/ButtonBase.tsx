import * as React from 'react';
import { ButtonProps } from './Button.types';
import { compose, ComposePreparedOptions } from '@fluentui/react-compose';
import { useButton } from './useButton';

export const ButtonBase = compose<'button', ButtonProps, ButtonProps, {}, {}>(
  (props: ButtonProps, ref: React.Ref<HTMLButtonElement>, composeOptions: ComposePreparedOptions<ButtonProps>) => {
    const { slots, slotProps } = useButton(props, composeOptions);

    return (
      <slots.root ref={ref} {...slotProps.root}>
        {props.loading && <slots.loader {...slotProps.loader} />}
        {props.icon && props.iconPosition !== 'after' && <slots.icon {...slotProps.icon} />}
        {!props.iconOnly && props.content && <slots.content {...slotProps.content} />}
        {props.icon && props.iconPosition === 'after' && <slots.icon {...slotProps.icon} />}
      </slots.root>
    );
  },
  {
    slots: {
      icon: 'span',
      content: 'span',
      loader: 'span',
    },
    displayName: 'ButtonBase',

    handledProps: [
      'circular',
      'content',
      'disabled',
      'fluid',
      'iconOnly',
      'iconPosition',
      'inverted',
      'loader',
      'loading',
      'primary',
      'secondary',
      'size',
    ],
  },
);

ButtonBase.defaultProps = {
  as: 'button',
};
