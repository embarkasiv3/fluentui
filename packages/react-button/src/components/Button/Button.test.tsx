import * as React from 'react';
import * as path from 'path';
import { isConformant } from '@fluentui/react-conformance';
import { Button as ButtonComponent } from './Button';
import * as renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';
import { MergeStylesProvider } from '@fluentui/react-theme-provider';
import { ButtonProps } from './Button.types';

/** Use merge-styles provider to ensure styles show up in snapshots. */
const Button = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => (
  <MergeStylesProvider>
    <ButtonComponent {...props} ref={ref} />
  </MergeStylesProvider>
));

describe('Button', () => {
  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  isConformant({
    componentPath: path.join(__dirname, 'Button.tsx'),
    Component: ButtonComponent,
    displayName: 'Button',
    disabledTests: ['has-docblock', 'as-renders-html', 'as-passes-as-value', 'as-renders-react-class', 'as-renders-fc'],
  });

  /**
   * Note: see more visual regression tests for Button in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<Button>Default button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders anchor when href prop is provided', () => {
    const component = renderer.create(<Button href="https://www.bing.com">Default button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can be focused', () => {
    const rootRef = React.createRef<HTMLButtonElement>();

    wrapper = mount(<Button ref={rootRef}>Focus me</Button>);

    expect(typeof rootRef.current).toEqual('object');
    expect(document.activeElement).not.toEqual(rootRef.current);

    rootRef.current?.focus();

    expect(document.activeElement).toEqual(rootRef.current);
  });
});
