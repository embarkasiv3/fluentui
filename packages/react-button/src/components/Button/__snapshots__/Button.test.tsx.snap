// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a default state 1`] = `
<button
  className=
      ms-Button
      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        align-items: center;
        background: var(--button-background, white);
        border-bottom-left-radius: var(--button-borderBottomLeftRadius, var(--button-borderRadius));
        border-bottom-right-radius: var(--button-borderBottomRightRadius, var(--button-borderRadius));
        border-bottom-width: var(--button-borderBottomWidth, var(--button-borderWidth));
        border-color: var(--button-borderColor);
        border-left-width: var(--button-borderLeftWidth, var(--button-borderWidth));
        border-right-width: var(--button-borderRightWidth, var(--button-borderWidth));
        border-style: solid;
        border-top-left-radius: var(--button-borderTopLeftRadius, var(--button-borderRadius));
        border-top-right-radius: var(--button-borderTopRightRadius, var(--button-borderRadius));
        border-top-width: var(--button-borderTopWidth, var(--button-borderWidth));
        box-shadow: var(--button-boxShadow);
        box-sizing: border-box;
        color: var(--button-contentColor);
        cursor: pointer;
        display: inline-flex;
        font-family: var(--button-fontFamily);
        font-size: var(--button-fontSize);
        font-weight: var(--button-fontWeight);
        height: var(--button-height);
        justify-content: center;
        max-width: var(--button-maxWidth);
        min-height: var(--button-minHeight);
        min-width: var(--button-minWidth);
        outline: none;
        padding-bottom: var(--button-paddingBottom);
        padding-left: var(--button-paddingLeft);
        padding-right: var(--button-paddingRight);
        padding-top: var(--button-paddingTop);
        position: relative;
        text-decoration: none;
        transition: var(--button-transition);
        user-select: none;
        vertical-align: middle;
        white-space: var(--button-whiteSpace);
        width: var(--button-width);
      }
      .ms-Fabric--isFocusVisible &:focus::after {
        border-bottom-left-radius: var(--button-borderBottomLeftRadius, var(--button-borderRadius));
        border-bottom-right-radius: var(--button-borderBottomRightRadius, var(--button-borderRadius));
        border-color: var(--button-focusColor, black);
        border-style: solid;
        border-top-left-radius: var(--button-borderTopLeftRadius, var(--button-borderRadius));
        border-top-right-radius: var(--button-borderTopRightRadius, var(--button-borderRadius));
        border-width: var(--button-focusWidth, 1.6px);
        bottom: -1px;
        box-shadow: 0 0 0 var(--button-focusInnerWidth, 1px) var(--button-focusInnerColor, white) inset;
        content: "";
        left: -1px;
        position: absolute;
        right: -1px;
        top: -1px;
        z-index: 1;
      }
      & > *:not(:first-child) {
        margin-left: var(--button-contentGap);
      }
      &:hover {
        background: var(--button-hovered-background, var(--button-background));
        border-color: var(--button-hovered-borderColor, var(--button-borderColor));
        box-shadow: var(--button-hovered-boxShadow, var(--button-boxShadow));
        color: var(--button-hovered-contentColor, var(--button-contentColor));
      }
      &:hover .ms-Button-icon {
        color: var(--button-hovered-iconColor, var(--button-iconColor));
      }
      &:active {
        background: var(--button-pressed-background, var(--button-hovered-background));
        border-color: var(--button-pressed-borderColor, var(--button-hovered-borderColor, var(--button-borderColor)));
        box-shadow: var(--button-pressed-boxShadow, var(--button-hovered-boxShadow, var(--button-boxShadow)));
        color: var(--button-pressed-contentColor, var(--button-hovered-contentColor, var(--button-contentColor)));
        transform: var(--button-pressed-transform);
        transition: var(--button-pressed-transition);
      }
      &:active .ms-Button-icon {
        color: var(--button-pressed-iconColor, var(--button-iconColor));
      }
      &[aria-disabled=true] {
        background-color: var(--button-disabled-background);
        border-color: var(--button-disabled-borderColor);
        box-shadow: var(--button-disabled-boxShadow);
        color: var(--button-disabled-contentColor);
        opacity: var(--button-disabled-opacity);
        pointer-events: none;
      }
      &[aria-disabled=true] .ms-Button-icon {
        color: var(--button-disabled-iconColor);
      }
>
  <span
    className=

        {
          line-height: 1;
        }
  >
    Default button
  </span>
</button>
`;

exports[`Button renders anchor when href prop is provided 1`] = `
<a
  className=
      ms-Button
      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        align-items: center;
        background: var(--button-background, white);
        border-bottom-left-radius: var(--button-borderBottomLeftRadius, var(--button-borderRadius));
        border-bottom-right-radius: var(--button-borderBottomRightRadius, var(--button-borderRadius));
        border-bottom-width: var(--button-borderBottomWidth, var(--button-borderWidth));
        border-color: var(--button-borderColor);
        border-left-width: var(--button-borderLeftWidth, var(--button-borderWidth));
        border-right-width: var(--button-borderRightWidth, var(--button-borderWidth));
        border-style: solid;
        border-top-left-radius: var(--button-borderTopLeftRadius, var(--button-borderRadius));
        border-top-right-radius: var(--button-borderTopRightRadius, var(--button-borderRadius));
        border-top-width: var(--button-borderTopWidth, var(--button-borderWidth));
        box-shadow: var(--button-boxShadow);
        box-sizing: border-box;
        color: var(--button-contentColor);
        cursor: pointer;
        display: inline-flex;
        font-family: var(--button-fontFamily);
        font-size: var(--button-fontSize);
        font-weight: var(--button-fontWeight);
        height: var(--button-height);
        justify-content: center;
        max-width: var(--button-maxWidth);
        min-height: var(--button-minHeight);
        min-width: var(--button-minWidth);
        outline: none;
        padding-bottom: var(--button-paddingBottom);
        padding-left: var(--button-paddingLeft);
        padding-right: var(--button-paddingRight);
        padding-top: var(--button-paddingTop);
        position: relative;
        text-decoration: none;
        transition: var(--button-transition);
        user-select: none;
        vertical-align: middle;
        white-space: var(--button-whiteSpace);
        width: var(--button-width);
      }
      .ms-Fabric--isFocusVisible &:focus::after {
        border-bottom-left-radius: var(--button-borderBottomLeftRadius, var(--button-borderRadius));
        border-bottom-right-radius: var(--button-borderBottomRightRadius, var(--button-borderRadius));
        border-color: var(--button-focusColor, black);
        border-style: solid;
        border-top-left-radius: var(--button-borderTopLeftRadius, var(--button-borderRadius));
        border-top-right-radius: var(--button-borderTopRightRadius, var(--button-borderRadius));
        border-width: var(--button-focusWidth, 1.6px);
        bottom: -1px;
        box-shadow: 0 0 0 var(--button-focusInnerWidth, 1px) var(--button-focusInnerColor, white) inset;
        content: "";
        left: -1px;
        position: absolute;
        right: -1px;
        top: -1px;
        z-index: 1;
      }
      & > *:not(:first-child) {
        margin-left: var(--button-contentGap);
      }
      &:hover {
        background: var(--button-hovered-background, var(--button-background));
        border-color: var(--button-hovered-borderColor, var(--button-borderColor));
        box-shadow: var(--button-hovered-boxShadow, var(--button-boxShadow));
        color: var(--button-hovered-contentColor, var(--button-contentColor));
      }
      &:hover .ms-Button-icon {
        color: var(--button-hovered-iconColor, var(--button-iconColor));
      }
      &:active {
        background: var(--button-pressed-background, var(--button-hovered-background));
        border-color: var(--button-pressed-borderColor, var(--button-hovered-borderColor, var(--button-borderColor)));
        box-shadow: var(--button-pressed-boxShadow, var(--button-hovered-boxShadow, var(--button-boxShadow)));
        color: var(--button-pressed-contentColor, var(--button-hovered-contentColor, var(--button-contentColor)));
        transform: var(--button-pressed-transform);
        transition: var(--button-pressed-transition);
      }
      &:active .ms-Button-icon {
        color: var(--button-pressed-iconColor, var(--button-iconColor));
      }
      &[aria-disabled=true] {
        background-color: var(--button-disabled-background);
        border-color: var(--button-disabled-borderColor);
        box-shadow: var(--button-disabled-boxShadow);
        color: var(--button-disabled-contentColor);
        opacity: var(--button-disabled-opacity);
        pointer-events: none;
      }
      &[aria-disabled=true] .ms-Button-icon {
        color: var(--button-disabled-iconColor);
      }
  href="https://www.bing.com"
  role="button"
>
  <span
    className=

        {
          line-height: 1;
        }
  >
    Default button
  </span>
</a>
`;
